### ISSUE TITLE
[except.spec] Making the program ill-formed when the referred to function in expression E is the same as for which the predicate noexcept operator is being evaluated

### ISSUE THREAD

**Issue by ranaanoop** [cplusplus/CWG#152]
Full name of submitter (unless configured in github; will be published with the issue): Anoop Rana

Reference (section label): [expect.spec#6]

Link to reflector thread (if any):

Issue description: Currently it is not clear whether the following program is well-formed or ill-formed and according to which clause from the standard. To remedy this or make this clearly ill-formed, i think that we should add explicitly add a statement(not as a bullet point) in [expect.spec#6] to say something like "in case the expression `E` is a function call, the referred to function(by `E`) must not be the same as the function for which the predicate is being evaluated". 

```
struct C {
    constexpr static bool f() noexcept(!noexcept(C::f())) 
    {
        return true;
    }
};
```
We observe [implementation divergence](here). 

<h3>Suggested resolution:</h3>

Add a statement(not as a bullet point but a separate statement) in [expect.spec#6](https://timsong-cpp.github.io/cppwp/n4861/except.spec#6) saying something along the lines of: 

> In case the expression `E` is a function call, the referred to function(by `E`) must not be the same as the function for which the predicate is being evaluated. 

(end suggestion) 

This means that since `E` is `C::f()` and the function referred to by `E` is `f` which is the same function for which the predicate is being evaluated so that the result of the predicate can be used in the noexcept-specifier, so the program is ill-formed by using the suggested resolution. 

<h4>SideNote</h4>

I'm confident that there are better  ways to phrase the suggested resolution, but due to me not being a native english speaker I was unable to phrase it in a more concise manner. Even then, the basic point that i'm making remains that same which is that the function referred to by `E` must not be the same as the function for which `noexecpt-specifer` predicate we're evaluating.

**Comment by languagelawyer** [comment]
How about mutually-recursive references?

**Comment by ranaanoop** [comment]
@languagelawyer I suppose you're suggesting a better term/phrase, right?

**Comment by frederick-vs-ja** [comment]
I think it's possible to define the dependency relationship of _noexcept-specifier_ s inside a class, and require that there's no cycle.

**Comment by xmh0511** [comment]
I think Clang has the right behavior, [dcl.fct] p12 says
> The return type, the parameter-type-list, the ref-qualifier, the cv-qualifier-seq, and **the exception specification**, but not the default arguments ([dcl.fct.default]) or the trailing requires-clause ([dcl.decl]), **are part of the function type**.

[except.spec] p6.1 says
> E is a function call whose postfix-expression **has a function type**, or a pointer-to-function type, with a potentially-throwing exception specification

In your example, the function type cannot be determined in the context of `noexcept(...)` because it is a part of the function type that is being determined. I think we should let the relevant rule clearer.   

***
However, things turn in complex if we introduce a template here
````cpp
struct C {
    template<class T>
    constexpr static bool f(T) noexcept(!noexcept(f(T{}))){
        return true;
    }
};
int main(){
    C::f(0); // cause the error
}
````

**Comment by ranaanoop** [comment]
Or the following wording could be added to make the program ill-formed: 

> While evaluating the result of a `noexcept operator` predicate nested inside a `noexcept-specifier` for a function `F`, the operand of the `noexcept operand` will not refer `F`.

**Comment by xmh0511** [comment]
> Or the following wording could be added to make the program ill-formed:
> 
> > While evaluating the result of a `noexcept operator` predicate nested inside a `noexcept-specifier` for a function `F`, the operand of the `noexcept operand` will not recursively refer `F`.

No, I think `constexpr static bool f1() noexcept( sizeof(noexcept(C::f1())) );` should work.  `noexcept(C::f1())` does refer to `F` however the result of the predicate does not rely on the result of the innermost `noexcept(C::f1())`.
